### PR TITLE
Implemented resume_gateway_url.

### PR DESCRIPTION
**Pull request by Quahu**
Implemented the new `resume_gateway_url`.

I added two unrelated to my changes TODOs in `DiscordSocketApiClient.ConnectAsync()` for logic that should probably be adjusted at some point, but I can remove them if you'd like me to.

### CODE REVIEW

**Review by csmir — APPROVED** [review]

**Review by MrCakeSlayer — APPROVED** [review]

File: src/Discord.Net.WebSocket/DiscordSocketApiClient.cs [hunk 1]
```diff
@@ -191,24 +211,41 @@ internal override async Task ConnectInternalAsync()
...
                     WebSocketClient.SetCancelToken(_connectCancelToken.Token);
 
-                if (!_isExplicitUrl)
+                string gatewayUrl;
+                if (_resumeGatewayUrl == null)
+                {
+                    if (!_isExplicitUrl)
+                    {
+                        // TODO: 'GetGatewayAsync' -> 'GetBotGatewayAsync', but it could just be hardcoded to 'wss://gateway.discord.gg/'
+                        var gatewayResponse = await GetGatewayAsync().ConfigureAwait(false);
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]:
```suggestion
                        var gatewayResponse = await GetBotGatewayAsync().ConfigureAwait(false);
```

From DV8:
> - new sessions go to /bot/gateway
> - resumable session to to resume_url
> - if a resume fails, you need a new session

File: src/Discord.Net.WebSocket/DiscordSocketApiClient.cs [hunk 2]
```diff
@@ -191,24 +211,41 @@ internal override async Task ConnectInternalAsync()
...
 
                 ConnectionState = ConnectionState.Connected;
             }
             catch
             {
                 if (!_isExplicitUrl)
+                {
+                    // TODO: '_gatewayUrl = null' doesn't do anything, it's never null checked
                     _gatewayUrl = null; //Uncache in case the gateway url changed
+                }
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]:
```suggestion
```

**Review by quinchs — CHANGES_REQUESTED** [review]

**Comment by quinchs** [comment]
@Quahu Your move with the suggestions

File: src/Discord.Net.WebSocket/DiscordSocketApiClient.cs (hunk 1 above)
Comment (Quahu) [reply src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]: It's the same thing, i.e. `wss://gateway.discord.gg/`.
I'd just hardcode it, which would also solve the issue of individual shards all re-requesting this value redundantly.
I adjusted the TODO to mention that.

Comment (quinchs) [reply src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]: Do you wanna code this in?

Comment (Quahu) [reply src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]: You mean hardcode it?

File: src/Discord.Net.WebSocket/DiscordSocketApiClient.cs [hunk 3]
```diff
@@ -191,24 +211,35 @@ internal override async Task ConnectInternalAsync()
...
                     WebSocketClient.SetCancelToken(_connectCancelToken.Token);
 
-                if (!_isExplicitUrl)
+                string gatewayUrl;
+                if (_resumeGatewayUrl == null)
+                {
+                    if (!_isExplicitUrl)
+                    {
+                        // TODO: pass this down from DiscordShardedClient, so that it's not requested separately for every single shard
+                        var gatewayResponse = await GetBotGatewayAsync().ConfigureAwait(false);
```
Comment (quinchs) [inline src/Discord.Net.WebSocket/DiscordSocketApiClient.cs]: Implement the TODO

**Review by quinchs — CHANGES_REQUESTED** [review]

**Review by quinchs — APPROVED** [review]

**Review by csmir — APPROVED** [review]
lgtm

**Pull request merged**